core/server/proxy/src/main/java/alluxio/proxy/s3/RangeFileInStream.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.proxy.s3;
+
+import alluxio.client.file.FileInStream;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * This class is use {@link FileInStream} underlying, and implement range read.
+ */
+public class RangeFileInStream extends InputStream {
+
+  private FileInStream mUnderlyingStream;
+  private long mUnderlyingLength;
+  private long mReadBytes;
+
+  private RangeFileInStream(FileInStream underlyingStream) {
+    mUnderlyingStream = underlyingStream;
+    mReadBytes = 0;
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (mReadBytes >= mUnderlyingLength) {
+      return -1;
+    }
+
+    int b = mUnderlyingStream.read();
+    if (b != -1) {
+      mReadBytes++;
+    }
+    return b;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (mReadBytes == mUnderlyingLength) {
```

Comment:
as a defensive measure, I would use `>= mUnderlyingLength` in the case the logic (whether now or in the future) ever goes over the `mUnderlyingThreshold`